Second complete version reviewed

TheLokj opened the pull request:
Thanks for your previous recommandations !

I've modified the pipeline following your advices and I had to change few other things, to optimize it notably (by using Python script and SeqKit instead of just launching modules on each contig within a job running a bash command, which obviously made the pipeline slower and HPC managers furious ; more details in the commits descriptions). 

(The pipeline is currently working well on real data, so take your time it's not an emergency)

TheLokj committed:
First commit

TheLokj committed:
Correction on the .sif path in README.md

TheLokj committed:
Merge branch 'main' into dev

TheLokj committed:
Improve the code structuration

TheLokj committed:
Improve the code structuration

TheLokj committed:
Merge branch 'dev' of https://github.com/TheLokj/kingdom-investigator into dev

TheLokj committed:
Adding the second step of validation using CAT

TheLokj committed:
Adding the second step of validation using CAT

TheLokj committed:
Update README.md

TheLokj committed:
Update bin/parse_dmc_classification.py

Co-authored-by: Varsha Kale <34109092+Vkale1@users.noreply.github.com>

TheLokj committed:
Replace sed and awk script by SeqKit in get_contigs_info.nf and save_contigs_per_kingdom.nf

Update README.md and add the decision rule table

Change decision rules

Add Python containers for Pythin script

Update the metadata.json structure and make the save_metadata script more human-readable

Remove the duplicate CAT container

Simplify some scripts

TheLokj committed:
Merge branch 'dev' of https://github.com/TheLokj/kingdom-investigator into dev

TheLokj committed:
Remove conflicts

TheLokj committed:
Upload the current DAG

TheLokj committed:
Add the work in progress DAG in the README.md

TheLokj committed:
parse_dmc_classification.py correction

TheLokj committed:
Optimize the pipeline by launching only one time the decisions and save_contigs modules {launched for each contig previously)

Rename step to stage and change the fasta names to lower case ones

TheLokj committed:
Add the files related to the previous commit

TheLokj committed:
Change seqkit from -p to -f to avoid too many args limit

TheLokj committed:
Replace process ressource labels with the nf-core naming convention

TheLokj committed:
Replace process ressource labels with the nf-core naming convention

TheLokj committed:
Update CAT ressource allocation to make it compatible with real use case and avoid a OOM kill

TheLokj commented:
> Thanks for your previous recommandations !
> 
> I've modified the pipeline following your advices and I had to change few other things, to optimize it notably (by using Python script and SeqKit instead of just launching modules on each contig within a job running a bash command, which obviously made the pipeline slower and HPC managers furious ; more details in the commits descriptions).
> 
> (The pipeline is currently working well on real data, so take your time it's not an emergency)

(Oh and don't worry about the conflicts I'll resolve these ones after, need to improve myself in Git)

vagkaratzas commented on modules/save_contigs_per_kingdom.nf:
```diff
@@ -1,17 +1,16 @@
-  val contigsPath
-  each result
+  path contigsPath
+  each ids
```
Just a minor comment here. In my pipelines when I don't want to run a process per file (because it will spawn lots of jobs in the cluster), I put the files in a folder and feed that folder path to my process instead. Then I call a python script from the module that parses/executes the files one by one, putting their results in one main folder, again which will be the output of my module. I think that would be equal functionality to how you use 'each'? But not sure since I've never used it. If it works as expected here, leave it as it is of course. Just make sure you get the resutl you would expect ;)

TheLokj replied:
I didn't thought about this possibiliy thanks, it can be really useful !
The `each ids` allows me here to run the module for each path in the input path channel individually. As each path is a file representing a predicted kingdom and containing the IDs of associated contigs, it can only be launched maximum 4 times (eukaryotes, prokaryotes, unknwon, requiringCATvalidation), so it's not a lot, but your idea could definitively be useful in larger cases ! :)

vagkaratzas reviewed (commented):
Thanks for implementing all the changes I asked for. I would opt for a smaller example if possible to run faster as test case (there are 100 items now, but they used to be 10 I think?). Also, another advice is to make sure you put in the publishDir only the final files you want to output to the user in the end, for storage purposes. For example, I am not sure we want the ../firstStage/classifiedAsEukaryote.fa. Even if so, maybe a file with just the name ">ERZ1463014.522 NODE_522_length_15319_cov_43.265985" would be enough instead of the whole fasta? Just an example there, just make sure.

TheLokj commented:
> Thanks for implementing all the changes I asked for. I would opt for a smaller example if possible to run faster as test case (there are 100 items now, but they used to be 10 I think?). Also, another advice is to make sure you put in the publishDir only the final files you want to output to the user in the end, for storage purposes. For example, I am not sure we want the ../firstStage/classifiedAsEukaryote.fa. Even if so, maybe a file with just the name ">ERZ1463014.522 NODE_522_length_15319_cov_43.265985" would be enough instead of the whole fasta? Just an example there, just make sure.

I'll reduce the test set you're right. And yes, I did that at the beggining of the pipeline writing, but when I think about it now after your message, in any case the `summary.tsv` contains the result of each step and the ids of the associated contigs, so it's a useless waste of space. Thank you!

TheLokj committed:
Remove the first stage multifastas saved in the publishDir

Change the first stage to only save the contigs requiring CAT validation in the working directory in order to optimise storage use

TheLokj committed:
Add a smaller test dataset

TheLokj committed:
Rename the pipeline

Change the decision rules

Reduce the test dataset

TheLokj committed:
Upload the new test dataset

TheLokj committed:
Delete figures/decisions.png

TheLokj committed:
Update decisions.png
